contrib/natss/pkg/dispatcher/dispatcher/dispatcher.go

```diff
@@ -83,10 +108,66 @@ func createReceiverFunction(s *SubscriptionsSupervisor, logger *zap.SugaredLogge
 }
 
 func (s *SubscriptionsSupervisor) Start(stopCh <-chan struct{}) error {
+	// Starting Connect to establish connection with NATS
+	go s.Connect(stopCh)
+	// Trigger Connect to establish connection with NATS
+	s.connect <- struct{}{}
 	s.receiver.Start(stopCh)
 	return nil
 }
 
+func (s *SubscriptionsSupervisor) connectWithRetry(stopCh <-chan struct{}) {
+	// re-attempting evey 60 seconds until the connection is established.
+	ticker := time.NewTicker(60 * time.Second)
+	defer ticker.Stop()
+	for {
+		nConn, err := stanutil.Connect(clusterchannelprovisioner.ClusterId, clientID, s.natssURL, s.logger.Sugar())
+		if err == nil {
+			// Locking here in order to reduce time in locked state
+			s.natssConnMux.Lock()
+			s.natssConn = nConn
+			s.natssConnInProgress = false
+			s.natssConnMux.Unlock()
+			return
+		}
+		s.logger.Sugar().Errorf("Connect() failed with error: %+v, retrying in 60 seconds", err)
+		select {
+		case <-ticker.C:
+			continue
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// Connect is called for initial connection as well as after every disconnect
+func (s *SubscriptionsSupervisor) Connect(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-s.connect:
+			if s.natssConn == nil && !s.natssConnInProgress {
+				// Case for Initial connect, setting up InProgress to true to prevent recursion
+				s.natssConnMux.Lock()
+				s.natssConnInProgress = true
+				s.natssConnMux.Unlock()
+				go s.connectWithRetry(stopCh)
+				continue
+			}
+			if !(*s.natssConn).NatsConn().IsConnected() &&
```

Comment:
If `s.natssConn` is `nil` and `s.natssConnInProgress` is `true`, will this panic (because it tries to dereference `s.natssConn`)?

Reply:
provided a protection against this corner case.